Usability fixes

marcelveldt opened the pull request:
**BREAKING CHANGE**
This is a game changer regarding to usability. This will make sure that sensors are created with sensible names (and values) and properly attached to a device class.

More advanced sensors are disabled by default. Powerusers can enable them if they really need them. Sensors in notification class are translated to binary sensors.

This is a breaking change because all entities will have a new unique ID, to make sure we can have multiple sensors based from the same Z-Wave Values. 
Now that we're still in pre-beta phase this is acceptable.

Related issue: #69

marcelveldt committed:
prefer text value

For list values, prefer the label as it is much more user friendly.

marcelveldt committed:
Cleanup discovery and sensors

marcelveldt committed:
Allow multiple entities with same value

marcelveldt committed:
also add plain sensors for backwards compatability

the plain sensors as they existed in old implementation are added by default disabled

marcelveldt committed:
some cleanup

marcelveldt committed:
prefer short devicename from metadata

using the shortened devicename from metadata gives a less cluttered default setup

marcelveldt committed:
disable legacy sensors

marcelveldt committed:
make sure we have unique device id's

OZW can have multiple instances, we did not cover that

marcelveldt committed:
only add generic sensor if we did not add it manually

marcelveldt commented:
<img width="344" alt="Schermafbeelding 2020-04-22 om 21 29 47" src="https://user-images.githubusercontent.com/6389780/80025108-67f6f500-84e0-11ea-8117-191ff432c664.png">

Example of a device with auto generated entities based on Notifications CommandClass.
Legacy and advanced sensors are available as disabled.

marcelveldt commented:
The tests will be fixed

MartinHjelmare commented on custom_components/zwave_mqtt/__init__.py:
```diff
@@ -212,6 +215,14 @@ def async_value_removed(value):
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Unload a config entry."""
+    # cleanup device registry (which will in turn clean entity registry)
+    dev_registry = await get_dev_reg(hass)
+    devices_to_remove = [
+        item.id for item in devices_for_config_entry(dev_registry, entry.unique_id)
+    ]
+    for dev_id in devices_to_remove:
+        dev_registry.async_remove_device(dev_id)
```
We shouldn't do this. The core is responsible for cleaning the device registry on config entry removal.

marcelveldt replied:
That's what I was discussing on discord. Accidentally left the code in ;-)
Anyways, with this present, devices are cleaned up, without it, they won't get cleaned out.

MartinHjelmare replied:
Let's remove it and look for the bug in the core.

marcelveldt replied:
fixed

MartinHjelmare commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -16,11 +30,41 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+            # we convert some of the Notification values into it's own binary sensor
+            # https://github.com/OpenZWave/open-zwave/blob/master/config/NotificationCCTypes.xml
+            for item in values.primary.value["List"]:
+                if values.primary.index == 6 and item["Value"] == 22:
```
Here we would like to use enums or similar from openzwavemqtt to avoid having to use the integers directly.

marcelveldt replied:
Yes, but that will be another PR. I am going to create constants and enums in the main lib.

marcelveldt replied:
added comment in the code to as reminder

MartinHjelmare commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -36,3 +80,101 @@ class ZWaveBinarySensor(ZWaveDeviceEntity, BinarySensorDevice):
+        return self.values.primary.value["Selected"] != "Clear"
+
+    @property
+    def state_attributes(self):
```
```suggestion
    def device_state_attributes(self):
```

marcelveldt replied:
Hmmm, in that case it needs more work as the base entity already has the device_state_attributes. I was hoping to not override that. What's the big difference between the two anyway ?

MartinHjelmare replied:
Each base entity integration, eg `components/binary_sensor/__init__.py`, implements `state_attributes`. This should not be overridden in the platforms. The platforms can implement `device_state_attributes` which is merged into `state_attributes` when updating state.

marcelveldt replied:
fixed

MartinHjelmare commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -36,3 +80,101 @@ class ZWaveBinarySensor(ZWaveDeviceEntity, BinarySensorDevice):
+    @property
+    def device_class(self):
+        """Return the class of this device, from component DEVICE_CLASSES."""
+        if self.values.primary.index == 1:
```
Break out a module level constant dict that maps index to device class, and look that dict up here.

marcelveldt replied:
done

MartinHjelmare commented on custom_components/zwave_mqtt/entity.py:
```diff
@@ -202,12 +203,12 @@ def device_state_attributes(self):
     def unique_id(self):
         """Return the unique_id of the entity."""
-        return self.values.unique_id
+        return f"{DOMAIN}.{self.platform.domain}.{self.values.values_id}"
```
The entity registry is aware of both integration domain and platform domain. We can remove those domains here.

marcelveldt replied:
Sure, I tried that before and I was getting unique_id errors without the platform.domain in the unique_id. Remember that we have sensors based on same values set in 2 different platforms (binary_sensor and sensor)

MartinHjelmare replied:
Unique ids only have to be unique per platform and integration. Only if there is more than one type of entity within the same platform for a device, do we need to add the entity type to the unique id. This is common in binary and sensor platforms that may spawn multiple entities for the same device.

marcelveldt replied:
You are right, I got on the wrong track with this because of those "platform does not generate unique id's" error but that is related to the reloading of the config entry and has has nothing to do with this PR. Reverted to the unique ID of the primary OZW Value again,

MartinHjelmare commented on custom_components/zwave_mqtt/sensor.py:
```diff
@@ -62,31 +105,23 @@ def unit_of_measurement(self):
-class ZWaveBatterySensor(ZWaveSensor):
-    """Representation of a Z-Wave battery sensor."""
+    @property
+    def state_attributes(self):
```
```suggestion
    def device_state_attributes(self):
```

marcelveldt replied:
fixed

marcelveldt committed:
cleanup device registry is handled by core

marcelveldt committed:
bring back instance event

instance_event callback got lost merging somehow?

marcelveldt committed:
use device_state_attributes

marcelveldt committed:
use mapping dict for the device classes

marcelveldt committed:
just use values_id for the entity unique id

domain/platform is already seperated by core

MartinHjelmare commented on custom_components/zwave_mqtt/binary_sensor.py:
```diff
@@ -107,33 +125,7 @@ def device_state_attributes(self):
-        if self.values.primary.index == 18:
-            return DEVICE_CLASS_GAS
-        return None
+        return DEVICE_CLASS_MAPPING.get(self.values.primary.index, None)
```
```suggestion
        return DEVICE_CLASS_MAPPING.get(self.values.primary.index)
```

marcelveldt committed:
fix binary_sensor name

also use the prettier shorter devicename as default

marcelveldt committed:
fix tests

marcelveldt committed:
fix tests

add test for the special binary sensors that are created
fix issues in the csv file, it contained 2 dead nodes and Selected_id was still missing (available in recent OZW version)

MartinHjelmare reviewed (approved):
Looks good!

marcelveldt commented:
Thanks for reviewing. I will merge this in. Let's get some feedback of users!
